File: amm/repl/api/src/main/java/ammonite/repl/api/FrontEnd.java
package ammonite.repl.api;

public abstract class FrontEnd {

  public FrontEnd() {

  }

  public static class Ammonite extends FrontEnd {
    public Ammonite() {

    }
  }

  public static Ammonite Ammonite() {
    return new Ammonite();
  }

  public static class JLineUnix extends FrontEnd {
    public JLineUnix() {

    }
  }

  public static JLineUnix JLineUnix() {
    return new JLineUnix();
  }

  public static class JLineWindows extends FrontEnd {
    public JLineWindows() {

    }
  }

  public static JLineWindows JLineWindows() {
    return new JLineWindows();
  }

}

File: amm/repl/api/src/main/scala/ammonite/repl/api/FrontEnd.scala
package ammonite.repl.api

import java.io.{InputStream, OutputStream}

import ammonite.util.{Colors, Res}

/** JLine interface */
trait FrontEnd {
  def width: Int
  def height: Int
  def action(
      input: InputStream,
      reader: java.io.Reader,
      output: OutputStream,
      prompt: String,
      colors: Colors,
      compilerComplete: (Int, String) => (Int, Seq[String], Seq[String]),
      history: IndexedSeq[String],
      addHistory: String => Unit,
      fileName: String
  ): Res[(String, Seq[String])]
}
